### PR TITLE
[node] Implement `matcher` config support for Middleware

### PR DESCRIPTION
**Pull request by TooTallNate**
Allows to specify a string or array of paths/globs for when a root-level Middleware should be invoked.

```typescript
export const config = {
  matcher: ['/about/:path*', '/dashboard/:path*'],
}

export default (req: Request) => {
  console.log(req.url);
  return new Response('hi');
};
```

Example deployment:

* Middleware is executed:
  * https://m-iotexsv6q-tootallnate.vercel.app/about
  * https://m-iotexsv6q-tootallnate.vercel.app/about/1
  * https://m-iotexsv6q-tootallnate.vercel.app/dashboard
  * https://m-iotexsv6q-tootallnate.vercel.app/dashboard/1
* Middleware _is not_ executed:
  * https://m-iotexsv6q-tootallnate.vercel.app/
  * https://m-iotexsv6q-tootallnate.vercel.app/other

### CODE REVIEW

File: packages/node/src/index.ts [hunk 1]
```diff
@@ -381,40 +382,48 @@ export const build: BuildV3 = async ({
...
+  if (isMiddleware) {
+    if (!isEdgeFunction) {
+      // Root-level middleware file can not have `export const config = { runtime: 'nodejs' }`
+      throw new Error(
+        `Middleware file can not be a Node.js Serverless Function`
+      );
+    }
+
+    // Middleware is a catch-all for all paths unless a `matcher` property is defined
+    const src = getRegExpFromMatchers(staticConfig?.matcher);
```
Comment (ijjk) [inline packages/node/src/index.ts:415]: Does the matcher need to consider the `entrypointPath` similar to the `outputName`?

**Review by cramforce — APPROVED** [review]

File: packages/cli/test/unit/commands/build.test.ts [hunk 2]
```diff
@@ -402,7 +402,72 @@ describe('build', () => {
...
+        ],
+      });
+
+      // `config.json` includes the "middlewarePath" route
+      const config = await fs.readJSON(join(output, 'config.json'));
+      expect(config).toMatchObject({
+        version: 3,
+        routes: [
+          {
+            src: '^\\/about(?:\\/((?:[^\\/#\\?]+?)(?:\\/(?:[^\\/#\\?]+?))*))?[\\/#\\?]?$|^\\/dashboard(?:\\/((?:[^\\/#\\?]+?)(?:\\/(?:[^\\/#\\?]+?))*))?[\\/#\\?]?$',
```
Comment (Schniz) [inline packages/cli/test/unit/commands/build.test.ts:467]: path-to-regexp generates such complex regexes 😮

File: packages/node/src/index.ts [hunk 3]
```diff
@@ -381,40 +382,48 @@ export const build: BuildV3 = async ({
...
+  }
+
+  // Add a `route` for Middleware
+  if (isMiddleware) {
+    if (!isEdgeFunction) {
+      // Root-level middleware file can not have `export const config = { runtime: 'nodejs' }`
+      throw new Error(
+        `Middleware file can not be a Node.js Serverless Function`
+      );
+    }
```
Comment (Schniz) [inline packages/node/src/index.ts:412]: **nit:** not related to this PR, but I wonder if the following would be simpler to understand, by not having `isEdgeFunction` and `isMiddleware` and inlining the usage as the conditionals are pretty straightforward:

```ts
const providedRuntime = AllowedRuntimes.parse(staticConfig.runtime); // pseudo code
if (config.middleware === true) {
  if (providedRuntime && providedRuntime !== 'experimental-edge') {
  ...
  }
}
```

**Review by Schniz — APPROVED** [review]
lgtm

File: packages/node/package.json [hunk 4]
```diff
@@ -56,6 +56,7 @@
     "content-type": "1.0.4",
     "cookie": "0.4.0",
     "etag": "1.8.1",
+    "path-to-regexp": "6.2.1",
```
Comment (styfle) [inline packages/node/package.json:59]: I wonder if we should update `@vercel/routing-tools` to match this latest version too (currently 6.1.0)

**Pull request merged**